### ISSUE TITLE
[Feature Request] Auto-batching inference server

### ISSUE THREAD

**Issue by dtsaras** [pytorch/rl#3462]
## Motivation

I am always disappointed when my GPUs are underutilized. I am currently working on having several agents operate in parallel on an MCTS tree. As the actors explore the tree independently, they need to query the model to select the next action. While each actor could technically keep its own copy of the model, this does not scale to a large number of parallel actors and can underutilize GPU resources.

Ideally, we would batch all model calls from the actors and serve them using one or more GPUs. In the LLM setting, this can be achieved by running a vLLM server: each actor sends requests to a shared server, which handles automatic batching, etc. While there are alternatives (e.g., Triton Inference Server or TorchServe, though TorchServe appears to be in limited maintenance), these are not ideal in a training loop.

The main issue is that there is no graceful way to update the parameters of the served model (and no tensordicts). With Triton, for example, you typically need to save a checkpoint, place it into the model repository, and then trigger a model load/reload. While workable, this is not elegant, and it can be slow enough that frequent updates become prohibitive.

## Solution

We can build our own more general inference server, potentially giving up some performance in exchange for a more training-friendly workflow. The properties we want are automatic batching, returning TensorDicts, a custom collate_fn, and the ability to integrate model weights with torch.distributed.

Queue
The queue is the primary interface for interacting with the served model. The inference worker consumes up to max_batch elements from the queue (when enough elements are available), or consumes all currently available elements once a time_limit is reached.

TensorDicts
The inference server should return the model’s TensorDict outputs to the appropriate actor. I’m not sure what the best way to implement this routing is.

collate_fn
Different users have different data formats, and variable-length data may require packing or padding. It would be nice to keep this flexible.

Weight sharing
We place the served model and the training model in a shared distributed group. During training, the user can broadcast updated weights efficiently to the inference model to minimize time spent waiting on synchronization.

## Alternatives

I considered Triton Inference Server and TorchServe, but neither satisfies the requirements above.

## Additional context

N/A

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)

**Comment by vmoens** [comment]
Is this the API you had in mind?
https://github.com/pytorch/rl/pull/3492

**Comment by dtsaras** [comment]
Yes! I think it would be a nice API such that environments can run easily in parallel, without having to worry about batching, especially if they can't be bathced nicely. It would also enable efficient large scale training. What are your thoughts?

**Comment by vmoens** [comment]
Yep we have async env so auto batching is the natural next step. Can't wait to see how we can hook it all up together!

**Comment by theap06** [comment]
@vmoens It has been a while since this feature has been opened. Could I take a crack at it?

**Comment by dtsaras** [comment]
> @vmoens It has been a while since this feature has been opened. Could I take a crack at it? 

I believe there are already some PRs for this. Maybe not merged yet

**Comment by vmoens** [comment]
See #3495 and associated stack
It's not very performant yet so I'd love to get your input on that feature if you want to help!
